plugins/askrene/graph.h

```diff
@@ -0,0 +1,165 @@
+#ifndef LIGHTNING_PLUGINS_ASKRENE_GRAPH_H
+#define LIGHTNING_PLUGINS_ASKRENE_GRAPH_H
+
+/* Solves a Minimum Cost Flow with arc selection costs and side constraints. */
+
+#include "config.h"
+#include <assert.h>
+#include <ccan/short_types/short_types.h>
+#include <ccan/tal/tal.h>
+
+#define INVALID_INDEX 0xffffffff
+
+/* A directed arc in a graph.
+ * It is a simple data object for typesafey. */
+struct arc {
+	/* arc's index */
+	u32 idx;
+};
+
+/* A node in a graph.
+ * It is a simple data object for typesafety. */
+struct node {
+	/* node's index */
+	u32 idx;
+};
+
+static inline struct arc arc_obj(u32 index)
+{
+	struct arc arc = {.idx = index};
+	return arc;
+}
+static inline struct node node_obj(u32 index)
+{
+	struct node node = {.idx = index};
+	return node;
+}
+
+/* A graph's topology. */
+struct graph {
+	/* Every arc emanates from a node, the tail.
+	 * The head of the arc is the tail of the dual. */
+	struct node *arc_tail;
+
+	/* Adjacency data for nodes. Used to move in a graph in the direction of
+	 * the arcs by looping over all arcs that exit a node.
+	 *
+	 * For every directed arc there is a dual in the opposite direction,
+	 * therefore we can use the same adjacency information to traverse in
+	 * the head to tails direction as well. */
+	struct arc *node_adjacency_next;
+	struct arc *node_adjacency_first;
+
+	size_t max_num_arcs, max_num_nodes;
+
+	/* Bit that must be flipped to obtain the dual of an arc. */
+	size_t arc_dual_bit;
+};
+
+//////////////////////////////////////////////////////////////////////////////
+
+static inline size_t graph_max_num_arcs(const struct graph *graph)
+{
+	return graph->max_num_arcs;
+}
+static inline size_t graph_max_num_nodes(const struct graph *graph)
+{
+	return graph->max_num_nodes;
+}
+
+/* Give me the dual of an arc. */
+static inline struct arc arc_dual(const struct graph *graph, struct arc arc)
+{
+	arc.idx ^= (1U << graph->arc_dual_bit);
+	return arc;
+}
+
+/* Is this arc a dual? */
+static inline bool arc_is_dual(const struct graph *graph, struct arc arc)
+{
+	return (arc.idx & (1U << graph->arc_dual_bit)) != 0;
+}
+
+/* Give me the node at the tail of an arc. */
+static inline struct node arc_tail(const struct graph *graph,
+				   const struct arc arc)
+{
+	assert(arc.idx < tal_count(graph->arc_tail));
```

Comment:
I suspect (but have not measured!) that this will introduce a perf regression.  Usually we don't care, but these are called so often that I suggest:

1. use graph_max_num_arcs()
2. Implement graph_check() which does these assertions, and call it in various places (e.g. end of constructor)

Reply:
I remember now the small performance tweak by avoiding call to `tal_count`. I can fix it.
But shouldn't we measure performance with release type of builds anyways? ie. compiled with -O2/3 and NDEBUG.

Reply:
Changed `tal_count` for `graph_max_num_arcs/nodes` in checks. d24e60a643e377a91163afb0271d42958b04d027